bpf/flows.c

```diff
@@ -95,59 +151,186 @@ static inline int fill_ip6hdr(struct ipv6hdr *ip, void *data_end, struct flow *f
         struct ipv6hdr *ip6 = (void *)eth + sizeof(*eth);
-        return fill_ip6hdr(ip6, data_end, flow);
+        return fill_ip6hdr(ip6, data_end, id, flags);
+    } else {
+        // TODO : Need to implement other specific ethertypes if needed
+        // For now other parts of flow id remain zero
+        memset (&(id->src_ip),0, sizeof(struct in6_addr));
+        memset (&(id->dst_ip),0, sizeof(struct in6_addr));
+        id->protocol = 0;
+        id->src_port = 0;
+        id->dst_port = 0;
     }
     return SUBMIT;
 }
 
-// parses flow information for a given direction (ingress/egress)
-static inline int flow_parse(struct __sk_buff *skb, u8 direction) {
 
+static inline int flow_monitor (struct __sk_buff *skb, u8 direction) {
     // If sampling is defined, will only parse 1 out of "sampling" flows
     if (sampling != 0 && (bpf_get_prandom_u32() % sampling) != 0) {
         return TC_ACT_OK;
     }
-
-    void *data = (void *)(long)skb->data;
     void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    flow_id id;
+    int rc = TC_ACT_OK;
+    u32 flags = 0;
+    flow_record *flow_event;
+    flow_aggregate *my_flow_aggregate;
 
-    struct flow *flow = bpf_ringbuf_reserve(&flows, sizeof(struct flow), 0);
-    if (!flow) {
+    u64 current_time = bpf_ktime_get_ns();
+
+    struct ethhdr *eth = data;
+    if (fill_ethhdr(eth, data_end, &id, &flags) == DISCARD) {
         return TC_ACT_OK;
     }
+    id.direction = direction;
+    if (direction == INGRESS) {
+        my_flow_aggregate = bpf_map_lookup_elem(&xflow_metric_map_ingress, &id);
+    } else {
+        my_flow_aggregate = bpf_map_lookup_elem(&xflow_metric_map_egress, &id);
+    }
 
-    struct ethhdr *eth = data;
-    if (fill_ethhdr(eth, data_end, flow) == DISCARD) {
-        bpf_ringbuf_discard(flow, 0);
+    if (my_flow_aggregate != NULL) {
+        // The key already exists in the map
+        if (flags & TCP_FIN_FLAG || flags & TCP_RST_FLAG) {
+            /* Need to evict the entry and send it via ring buffer */
+            flow_event = bpf_ringbuf_reserve(&flows, sizeof(flow_record), 0);
+            if (!flow_event) {
+                return rc;
+            }
+            //export_flow_id(&flow_event->flow_key, id, direction);
+            flow_event->id = id;
+            flow_event->metrics.flags = flags;
+            flow_event->metrics.packets = 1;
+            flow_event->metrics.bytes = skb->len;
+            flow_event->metrics.last_pkt_ts = current_time;
+            bpf_ringbuf_submit(flow_event, 0);
+            // Defer the deletion of the entry from the map to usespace since it evicts other CPU metrics
+        } else {
+            bool update = false;
+            my_flow_aggregate->packets += 1;
+            my_flow_aggregate->bytes += skb->len;
+            my_flow_aggregate->last_pkt_ts = current_time;
+            if (my_flow_aggregate->eth_protocol == 0) { // Ensures no other flow is residing
```

Comment:
Under which circumstance it could happen that `my_flow_aggregate != NULL` (it has been stored in the map) but the `eth_protocol` is 0 (as if no flow was stored)?

Reply:
This happens when a flow X from CPU 1 has already created the key and entry. Now, When flow Y from CPU2 hashes into the same bucket due to hash collision, this condition happens (i.e. flow_aggregate is not null)
In these cases, we still allow the flow Y to reside in the entry, and handle the collidedRecords in userspace.

Ofcourse, when flow X and flow Y are from the same CPU, we cannot let flow Y update the same bucket where flow X has active entries.

Reply:
I see. I must have misunderstood the meaning of `BPF_MAP_TYPE_PERCPU_HASH`. I thought it created a different map for each CPU so no possibilities of colliding two entries from the same CPU.